File: widgetssdk/src/main/java/com/glia/widgets/fileupload/domain/AddFileToAttachmentAndUploadUseCase.java
package com.glia.widgets.fileupload.domain;

import android.net.Uri;

import com.glia.androidsdk.Engagement;
import com.glia.androidsdk.engagement.EngagementFile;
import com.glia.widgets.core.engagement.GliaEngagementRepository;
import com.glia.widgets.core.engagement.exception.EngagementMissingException;
import com.glia.widgets.fileupload.FileAttachmentRepository;
import com.glia.widgets.fileupload.exception.RemoveBeforeReUploadingException;
import com.glia.widgets.fileupload.exception.SupportedFileCountExceededException;

public class AddFileToAttachmentAndUploadUseCase {
    private final static long SUPPORTED_FILE_COUNT = 25;

    private final GliaEngagementRepository gliaEngagementRepository;
    private final FileAttachmentRepository fileAttachmentRepository;

    public AddFileToAttachmentAndUploadUseCase(
            GliaEngagementRepository gliaEngagementRepository,
            FileAttachmentRepository fileAttachmentRepository
    ) {
        this.fileAttachmentRepository = fileAttachmentRepository;
        this.gliaEngagementRepository = gliaEngagementRepository;
    }

    public void execute(Uri fileUri, Listener listener) {
        if (fileAttachmentRepository.isFileAttached(fileUri)) {
            listener.onError(new RemoveBeforeReUploadingException());
        } else {
            onFileNotAttached(fileUri, listener);
        }
    }

    private void onFileNotAttached(Uri fileUri, Listener listener) {
        fileAttachmentRepository.attachFile(fileUri);
        if (hasNoOngoingEngagement()) {
            fileAttachmentRepository.setFileAttachmentEngagementMissing(fileUri);
            listener.onError(new EngagementMissingException());
        } else {
            onHasOngoingEngagement(fileUri, listener);
        }
    }

    private void onHasOngoingEngagement(Uri fileUri, Listener listener) {
        if (isSupportedFileCountExceeded()) {
            fileAttachmentRepository.setSupportedFileAttachmentCountExceeded(fileUri);
            listener.onError(new SupportedFileCountExceededException());
        } else {
            listener.onStarted();
            fileAttachmentRepository.uploadFile(fileUri, listener);
        }
    }

    private boolean isSupportedFileCountExceeded() {
        return fileAttachmentRepository.getAttachedFilesCount() > SUPPORTED_FILE_COUNT;
    }

    private boolean hasNoOngoingEngagement() {
        return !gliaEngagementRepository.hasOngoingEngagement();
    }

    public interface Listener {
        void onFinished();

        void onStarted();

        void onError(Exception ex);

        void onSecurityCheckStarted();

        void onSecurityCheckFinished(EngagementFile.ScanResult scanResult);
    }
}

File: widgetssdk/src/main/java/com/glia/widgets/internal/fileupload/domain/AddFileToAttachmentAndUploadUseCase.kt
package com.glia.widgets.internal.fileupload.domain

import com.glia.androidsdk.engagement.EngagementFile
import com.glia.widgets.internal.engagement.exception.EngagementMissingException
import com.glia.widgets.internal.fileupload.FileAttachmentRepository
import com.glia.widgets.internal.fileupload.exception.RemoveBeforeReUploadingException
import com.glia.widgets.internal.fileupload.exception.SupportedFileCountExceededException
import com.glia.widgets.internal.fileupload.exception.SupportedFileSizeExceededException
import com.glia.widgets.internal.fileupload.model.LocalAttachment
import com.glia.widgets.internal.secureconversations.domain.ManageSecureMessagingStatusUseCase
import com.glia.widgets.engagement.domain.IsQueueingOrLiveEngagementUseCase

internal class AddFileToAttachmentAndUploadUseCase(
    private val isQueueingOrLiveEngagementUseCase: IsQueueingOrLiveEngagementUseCase,
    private val fileAttachmentRepository: FileAttachmentRepository,
    private val manageSecureMessagingStatusUseCase: ManageSecureMessagingStatusUseCase
) {
    private val isSupportedFileCountExceeded: Boolean
        get() = fileAttachmentRepository.getAttachedFilesCount() > FileUploadLimitNotExceededObservableUseCase.FILE_UPLOAD_LIMIT

    operator fun invoke(file: LocalAttachment, listener: Listener) {
        if (fileAttachmentRepository.isFileAttached(file.uri)) {
            listener.onError(RemoveBeforeReUploadingException())
        } else {
            onFileNotAttached(file, listener)
        }
    }

    private fun onFileNotAttached(file: LocalAttachment, listener: Listener) {
        fileAttachmentRepository.attachFile(file)
        if (isQueueingOrLiveEngagementUseCase.hasOngoingLiveEngagement || manageSecureMessagingStatusUseCase.shouldBehaveAsSecureMessaging) {
            onHasOngoingOrSecureEngagement(file, listener)
        } else {
            fileAttachmentRepository.setFileAttachmentEngagementMissing(file.uri)
            listener.onError(EngagementMissingException())
        }
    }

    private fun onHasOngoingOrSecureEngagement(file: LocalAttachment, listener: Listener) {
        if (isSupportedFileCountExceeded) {
            fileAttachmentRepository.setSupportedFileAttachmentCountExceeded(file.uri)
            listener.onError(SupportedFileCountExceededException())
        } else if (isSupportedFileSizeExceeded(file)) {
            fileAttachmentRepository.setFileAttachmentTooLarge(file.uri)
            listener.onError(SupportedFileSizeExceededException())
        } else {
            listener.onStarted()
            fileAttachmentRepository.uploadFile(manageSecureMessagingStatusUseCase.shouldUseSecureMessagingEndpoints, file, listener)
        }
    }

    private fun isSupportedFileSizeExceeded(file: LocalAttachment): Boolean {
        return file.size >= SUPPORTED_FILE_SIZE
    }

    interface Listener {
        fun onFinished()
        fun onStarted()
        fun onError(ex: Exception)
        fun onSecurityCheckStarted()
        fun onSecurityCheckFinished(scanResult: EngagementFile.ScanResult?)
    }

    companion object {
        const val SUPPORTED_FILE_SIZE = 26214400L
    }
}
